substrate/frame/staking/stake-tracker/src/lib.rs

```diff
@@ -0,0 +1,756 @@
+//!
+//! It is important to ensure that the events are emitted from staking (i.e. the calls into
+//! [`OnStakingUpdate`]) *after* the staking ledger has been updated by the caller, since the new
+//! state will be fetched and used to update the sorted lists accordingly.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+pub use pallet::*;
+
+use frame_election_provider_support::SortedListProvider;
+use frame_support::{
+	defensive,
+	traits::{fungible::Inspect as FnInspect, Defensive, DefensiveSaturating},
+};
+use sp_runtime::traits::Zero;
+use sp_staking::{
+	currency_to_vote::CurrencyToVote, OnStakingUpdate, Stake, StakerStatus, StakingInterface,
+};
+use sp_std::{collections::btree_map::BTreeMap, vec, vec::Vec};
+
+#[cfg(test)]
+pub(crate) mod mock;
+#[cfg(test)]
+mod tests;
+
+pub(crate) const LOG_TARGET: &str = "runtime::stake-tracker";
+
+// syntactic sugar for logging.
+#[macro_export]
+macro_rules! log {
+	($level:tt, $patter:expr $(, $values:expr)* $(,)?) => {
+		log::$level!(
+			target: $crate::LOG_TARGET,
+			concat!("[{:?}] 📚 ", $patter), <frame_system::Pallet<T>>::block_number() $(, $values)*
+		)
+	};
+}
+
+/// The balance type of this pallet.
+pub type BalanceOf<T> = <<T as Config>::Staking as StakingInterface>::Balance;
+/// The account ID of this pallet.
+pub type AccountIdOf<T> = <T as frame_system::Config>::AccountId;
+
+/// Represents a stake imbalance to be applied to a staker's score.
+#[derive(Copy, Clone, Debug)]
+pub enum StakeImbalance<Balance> {
+	/// Represents the reduction of stake by `Balance`.
+	Negative(Balance),
+	/// Represents the increase of stake by `Balance`.
+	Positive(Balance),
+}
+
+impl<Balance: PartialOrd + DefensiveSaturating> StakeImbalance<Balance> {
+	fn from(prev_balance: Balance, new_balance: Balance) -> Self {
+		if prev_balance > new_balance {
+			StakeImbalance::Negative(prev_balance.defensive_saturating_sub(new_balance))
+		} else {
+			StakeImbalance::Positive(new_balance.defensive_saturating_sub(prev_balance))
+		}
+	}
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+	use crate::*;
+	use frame_election_provider_support::{ExtendedBalance, VoteWeight};
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::BlockNumberFor;
+
+	/// The current storage version.
+	const STORAGE_VERSION: StorageVersion = StorageVersion::new(0);
+
+	#[pallet::pallet]
+	#[pallet::storage_version(STORAGE_VERSION)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The stake balance.
+		type Currency: FnInspect<Self::AccountId, Balance = BalanceOf<Self>>;
+
+		/// The overarching event type.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// The staking interface.
+		type Staking: StakingInterface<AccountId = Self::AccountId>;
+
+		/// Something that provides an *always* sorted list of voters.
+		type VoterList: SortedListProvider<Self::AccountId, Score = VoteWeight>;
```

Comment:
```suggestion
		type VoterList: SortedListProvider<Self::AccountId, Score = VoteWeight>;

		type VoterUpdate: Get<VoterUpdateMode>;
```

Reply:
```
enum VoterUpdateMode {
	Lazy,
	Full,
}
```

Reply:
It doesn't. I have changed it to `u128` (extended balance) which simplified considerably the conversions between vote weight, balance, etc. The way I see it, at the end of the day the score of the target lists are a way to sort targets by approvals. The score itself is not used  for anything else than in the sorting process (yet, at least). So my reasoning to use u128 here is that 1) it should be safe for approvals stakes and 2) the vote weight/balance/ etc conversions are relatively minimal and it relies less on that call to total_issuance  (as if using balance as score).

In any case, I think that [this comment](https://github.com/paritytech/polkadot-sdk/pull/1933#discussion_r1583180875) is still relevant and we should address before SH.

Reply:
As noted elsewhere, please make an issue for removing all of this `VoteWeight`, `ExtendedBalance` and such in a future refactor. Especially if the staking parachain and AH are not connected, we cannot read into the total issuance so easily. 

We should always assume `Balance: Into<u128>`, and then do the calculation in `sp-npos-elections` in `U256`.

Reply:
I gave it a shot to add those changes to this PR but I think it's better to split in a separate PR. The tracking issue is here, needs to be done before the staking parachain https://github.com/paritytech/polkadot-sdk/issues/4725\.